core/arch/riscv/mm/core_mmu.c

```diff
@@ -0,0 +1,639 @@
+{
+	pte->entry = val;
+}
+
+static uint64_t core_mmu_entry_get(pte_t *pte)
+{
+	return pte->entry;
+}
+
+static inline bool core_mmu_entry_is_valid(pte_t *pte)
+{
+	return (pte->entry & PTE_V);
+}
+
+static inline bool core_mmu_entry_is_invalid(pte_t *pte)
+{
+	return !core_mmu_entry_is_valid(pte);
+}
+
+static inline bool core_mmu_entry_is_leaf(pte_t *pte)
+{
+	/* A leaf has one or more RWX bits set */
+	return (pte->entry & (PTE_R | PTE_W | PTE_X)) != 0;
+}
+
+static inline bool core_mmu_entry_is_branch(pte_t *pte)
+{
+	return !(core_mmu_entry_is_leaf(pte));
+}
+
+static inline unsigned long core_mmu_pte_create(unsigned long ppn, int type)
+{
+	return ((ppn << PTE_PPN_SHIFT) | PTE_V | type);
+}
+
+static inline unsigned long core_mmu_ptd_create(unsigned long ppn)
+{
+	return core_mmu_pte_create(ppn, PTE_V);
+}
+
+static unsigned long core_mmu_pte_ppn(pte_t *pte)
+{
+	return pte->entry >> PTE_PPN_SHIFT;
+}
+
+static unsigned long core_mmu_ppn(unsigned long addr)
+{
+	return (addr >> RISCV_PGSHIFT);
+}
+
+static unsigned long core_mmu_pte_pa(pte_t *pte)
+{
+	return (core_mmu_pte_ppn(pte) << RISCV_PGSHIFT);
+}
+
+static unsigned long core_mmu_pgt_to_satp(unsigned long asid, pgt_t *pgt)
+{
+	unsigned long satp = 0;
+	unsigned long pgt_ppn = ((unsigned long)pgt) >> RISCV_PGSHIFT;
+
+	assert(asid & g_asid == asid);
+	satp |= asid << RISCV_SATP_ASID_SHIFT;
+	satp |= (((unsigned long)RISCV_SATP_MODE) << RISCV_SATP_MODE_SHIFT);
+	satp |= pgt_ppn;
+
+	return satp;
+}
+
+static unsigned long pte_to_mattr(unsigned level __maybe_unused,
+				  unsigned long pte)
+{
+	unsigned long mattr = TEE_MATTR_SECURE;
+
+	mattr |= (pte & PTE_V) ? TEE_MATTR_VALID_BLOCK : 0;
+
+	if (pte & PTE_U) {
+		mattr |= (pte & PTE_R) ? TEE_MATTR_UR : 0;
+		mattr |= (pte & PTE_W) ? TEE_MATTR_UW : 0;
+		mattr |= (pte & PTE_X) ? TEE_MATTR_UX : 0;
+	} else {
+		mattr |= (pte & PTE_R) ? TEE_MATTR_PR : 0;
+		mattr |= (pte & PTE_W) ? TEE_MATTR_PW : 0;
+		mattr |= (pte & PTE_X) ? TEE_MATTR_PX : 0;
+	}
+
+	mattr |= (pte & PTE_G) ? TEE_MATTR_GLOBAL : 0;
+	mattr = (((PTE_R | PTE_W | PTE_X) == 0) & PTE_V) ? TEE_MATTR_TABLE : 0;
+
+	return mattr;
+}
+
+static unsigned long mattr_to_pte(unsigned level __maybe_unused, uint32_t attr)
+{
+	unsigned long pte = 0;
+
+	if (attr & TEE_MATTR_TABLE)
+		return PTE_V;
+
+	if (attr & TEE_MATTR_VALID_BLOCK)
+		pte |= PTE_V;
+
+	pte |= (attr & TEE_MATTR_UR) ? (PTE_R | PTE_U) : 0;
+	pte |= (attr & TEE_MATTR_UW) ? (PTE_W | PTE_U) : 0;
+	pte |= (attr & TEE_MATTR_UX) ? (PTE_X | PTE_U) : 0;
+
+	pte |= (attr & TEE_MATTR_PR) ? PTE_R : 0;
+	pte |= (attr & TEE_MATTR_PW) ? PTE_W : 0;
+	pte |= (attr & TEE_MATTR_PX) ? PTE_X : 0;
+
+	pte |= (attr & TEE_MATTR_GLOBAL) ? PTE_G : 0;
+
+	return pte;
+}
+
+static unsigned int core_mmu_pgt_idx(unsigned long addr, unsigned int level)
+{
+	unsigned int idx = addr >> (RISCV_PGLEVEL_BITS * level + RISCV_PGSHIFT);
+
+	return (idx & RISCV_MMU_VPN_MASK);
+}
+
+static inline unsigned int core_mmu_pgt_shift(unsigned int level)
```

Comment:
This function does exactly what `CORE_MMU_SHIFT_OF_LEVEL()` does. Is this function needed?

Reply:
Removed.